# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: mozglue/tests/TestBaseProfiler.cpp
# Commit: e582038fdae2
# Full Hash: e582038fdae25b304ea153cf10f498f78dd3cc71
# Author: Gerald Squelart <gsquelart@mozilla.com>
# Date: 2020-10-17 09:14:56
# Regressor Bug: 1670954
# File Overlap Count: 1
# Description:
#   Bug 1670954 - MarkerThreadId::MainThread() - r=gregtatum
#   
#   Since the main thread is almost always being profiled, this makes it easy to direct important markers there (e.g., FileIO markers from unregistered threads).
#   
#   Tech note: The `#include "BaseProfiler.h"` line was moved to BaseProfilerMarkersPrerequesites.h so that `MarkerThreadId::MainThread()` there can access `profiler_main_thread_id()`.
# ==============================================================================

diff -r a01db300cf22 -r e582038fdae2 mozglue/tests/TestBaseProfiler.cpp
--- a/mozglue/tests/TestBaseProfiler.cpp	Fri Oct 16 22:06:53 2020 +0000
+++ b/mozglue/tests/TestBaseProfiler.cpp	Fri Oct 16 22:07:20 2020 +0000
@@ -3816,6 +3816,45 @@
   printf("TestMarkerCategory done\n");
 }
 
+void TestMarkerThreadId() {
+  printf("TestMarkerThreadId...\n");
+
+  MOZ_RELEASE_ASSERT(MarkerThreadId{}.IsUnspecified());
+  MOZ_RELEASE_ASSERT(!MarkerThreadId::MainThread().IsUnspecified());
+  MOZ_RELEASE_ASSERT(!MarkerThreadId::CurrentThread().IsUnspecified());
+
+  MOZ_RELEASE_ASSERT(!MarkerThreadId{42}.IsUnspecified());
+  MOZ_RELEASE_ASSERT(MarkerThreadId{42}.ThreadId() == 42);
+
+  // We'll assume that this test runs in the main thread (which should be true
+  // when called from the `main` function).
+  MOZ_RELEASE_ASSERT(MarkerThreadId::MainThread().ThreadId() ==
+                     mozilla::baseprofiler::profiler_main_thread_id());
+
+  MOZ_RELEASE_ASSERT(MarkerThreadId::CurrentThread().ThreadId() ==
+                     mozilla::baseprofiler::profiler_current_thread_id());
+
+  MOZ_RELEASE_ASSERT(MarkerThreadId::CurrentThread().ThreadId() ==
+                     mozilla::baseprofiler::profiler_main_thread_id());
+
+  std::thread testThread([]() {
+    MOZ_RELEASE_ASSERT(!MarkerThreadId::MainThread().IsUnspecified());
+    MOZ_RELEASE_ASSERT(!MarkerThreadId::CurrentThread().IsUnspecified());
+
+    MOZ_RELEASE_ASSERT(MarkerThreadId::MainThread().ThreadId() ==
+                       mozilla::baseprofiler::profiler_main_thread_id());
+
+    MOZ_RELEASE_ASSERT(MarkerThreadId::CurrentThread().ThreadId() ==
+                       mozilla::baseprofiler::profiler_current_thread_id());
+
+    MOZ_RELEASE_ASSERT(MarkerThreadId::CurrentThread().ThreadId() !=
+                       mozilla::baseprofiler::profiler_main_thread_id());
+  });
+  testThread.join();
+
+  printf("TestMarkerThreadId done\n");
+}
+
 void TestMarkerNoPayload() {
   printf("TestMarkerNoPayload...\n");
 
@@ -3998,6 +4037,7 @@
   // ::SleepMilli(10000);
 
   TestMarkerCategory();
+  TestMarkerThreadId();
   TestMarkerNoPayload();
   TestUserMarker();
   TestPredefinedMarkers();
